hy/core/result_macros.py

```diff
@@ -0,0 +1,1491 @@
+        many(argument),
+        maybe(varargs("unpack-mapping", NASYM)))
+
+@pattern_macro(["fn", "fn/a"], [OPTIONAL_ANNOTATION, lambda_list, many(FORM)])
+def compile_function_lambda(compiler, expr, root, returns, params, body):
+    posonly, args, rest, kwonly, kwargs = params
+    has_annotations = returns is not None or any(
+        isinstance(param, tuple) and param[0] is not None
+        for param in (posonly or []) + args + kwonly + [rest, kwargs]
+    )
+    body = compiler._compile_branch(body)
+
+    # Compile to lambda if we can
+    if not has_annotations and not body.stmts and root != "fn/a":
+        args, ret = compile_lambda_list(compiler, params, Result())
+        return ret + asty.Lambda(expr, args=args, body=body.force_expr)
+
+    # Otherwise create a standard function
+    node = asty.AsyncFunctionDef if root == "fn/a" else asty.FunctionDef
+    name = compiler.get_anon_var()
+    ret = compile_function_node(compiler, expr, node, name, params, returns, body)
+
+    # return its name as the final expr
+    return ret + Result(expr=ret.temp_variables[0])
+
+@pattern_macro(["defn", "defn/a"], [OPTIONAL_ANNOTATION, SYM, lambda_list, many(FORM)])
+def compile_function_def(compiler, expr, root, returns, name, params, body):
+    node = asty.FunctionDef if root == "defn" else asty.AsyncFunctionDef
+    body = compiler._compile_branch(body)
+
+    return compile_function_node(
+        compiler, expr, node,
+        mangle(compiler._nonconst(name)), params, returns, body
+    )
+
+def compile_function_node(compiler, expr, node, name, params, returns, body):
+    ret = Result()
+    args, ret = compile_lambda_list(compiler, params, ret)
+
+    if body.expr:
+        body += asty.Return(body.expr, value=body.expr)
+
+    ret += node(
+        expr,
+        name=name,
+        args=args,
+        body=body.stmts or [asty.Pass(expr)],
+        decorator_list=[],
+        returns=compiler.compile(returns).force_expr if returns is not None else None,
+    )
+
+    ast_name = asty.Name(expr, id=name, ctx=ast.Load())
+    return ret + Result(temp_variables=[ast_name, ret.stmts[-1]])
+
+@pattern_macro("defmacro", [SYM | STR, lambda_list, many(FORM)])
+def compile_macro_def(compiler, expr, root, name, params, body):
+    _, _, rest, _, kwargs = params
+
+    if "." in name:
+        raise compiler._syntax_error(name, "periods are not allowed in macro names")
+    if rest == Symbol("*"):
+        raise compiler._syntax_error(rest, "macros cannot use '*'")
+    if kwargs is not None:
+        raise compiler._syntax_error(kwargs, "macros cannot use '#**'")
+
+    ret = Result() + compiler.compile(Expression([
+        Symbol("eval-and-compile"),
+        Expression([
+            Expression([
+                Symbol("hy.macros.macro"),
+                str(name),
+            ]),
+            Expression([
+                Symbol("fn"),
+                List([Symbol("hy-compiler")] + list(expr[2])),
```

Comment:
Can we call this `&compiler` to mimic the previous name `&name`, so that it's clear that it's a "special"/"magic" arg? Otherwise it's really confusing when looking at macros in e.g. hy/contrib/walk.hy trying to figure out where `hy-compiler` came from.

Reply:
I think it's polite to begin all variables we create with "hy_" (not to be confused with the mangling prefix, "hyx_") so the user knows that any variable name without this prefix is safe to use.

Reply:
`&hy_compiler`, perhaps?

Reply:
Blech, I realize `&hy_compiler` is also bad because it still won't have a "hy_" prefix right after the "hyx_"; the "XampersandX" will come first. Anything you'd like that starts with "hy_"?

Reply:
I'm not sure I understand your concern here; can we not just say `Symbol(mangle("&compiler"))` (or `Symbol(mangle("&hy-compiler"))`)?

Reply:
Sure we can. It's a matter of being nice to the user's namespace.

Reply:
To me, the leading `&` already ensures that we're not going to be stepping on uses'r names; there (was) already precedent for this with the old `&optional`, `&rest` etc. syntax, unless we want to completely move away from that.

Reply:
Aright, maybe I'm being a little too paranoid. I've rebased to use `&compiler`.